docs/specs/src/hwp_4_formal.md

```diff
@@ -0,0 +1,534 @@
+7. All counterparties now proceed to write the completed action to their source chain and publish its data to the DHT.
+8. The DHT authorities validate and store the action and entry data as normal.
+
+## Graph DHT: Formal State Model
+
+Holochain performs a topological transform on the set of the various agents' source chains into a content-addressable graph database (graph DHT or GDHT) sharded across many nodes who each function as authoritative sources for retrieving certain data.
+
+**Fundamental Write Constraint:** The DHT can never be "written" to directly. All data enters the DHT **only** by having been committed to an agent's source chain and then being **transformed** from validated local chain state into the elements (DHT operations) required for GDHT representation and lookup.
+
+**Structure of GDHT data:** The DHT is a content-addressable space where each piece of content is found at the address which is the hash of its content. In addition, any address in the DHT can have metadata attached to it. This metadata is not part of the content being hashed.
+
+*Note about hashing: Holochain uses 256-bit Blake2b hashes with the exception of one entry type, AgentPubKey, which is a 256-bit Ed25519 public key and uses `equal()` as its hash function. In other words, **the content of the AgentPubKey is identical to its hash**. This preserves content-addressability but also enables agent keys to function as self-proving identifiers for network transport and cryptographic functions like signing or encryption.*
+
+**DHT Addresses:** Both Actions and Entries from source chains can be retrieved from the DHT by either the ActionHash or EntryHash. The DHT `get()` function call returns a Record, a tuple containing the most relevant action/entry pair. Structurally, Actions "contain" their referenced entries so that pairing is obvious when a Record is retrieved by ActionHash. However, Actions are also attached as metadata at an EntryHash, and there could be many Actions which have created the same Entry content. A `get()` function called by EntryHash returns the oldest undeleted Action in the pair, while a `get_details()` function call on an EntryHash returns all of the Actions.
+
+**Agent Addresses & Agent Activity:** Technically an AgentPubKey functions as both a content address (which is never really used because performing a `get()` on the key just returns the key you used to get it) and a network address to send communications to that agents. But in addition to the content of the key stored on the DHT is metadata about that agent's chain activity. In other words, a `get_agent_activity()` request retrieves metadata about their chain records and chain status.
+
+Formally, the entire GDHT is represented as a set of 'basis hashes' $b_{c_x}$, or addresses where both content $c$ and metadata $m$ may be stored:
```

Comment:
I think these definitions should be explained. Especially $H$, which looks like it's the set of Actions, but its variable is named for its old name Headers.

Also, strictly speaking, the $c_x$ in the set of $d$ can also be "nothing", due to the ability to have "baseless links". So formally, I suggest defining $C = E \cup H \cup 0$, and explain that $0$ is introduced to represent absence of content at this basis, allowing for only metadata to be stored. Or something like that.

Reply:
> it's just that sometimes the data $c$ isn't stored at the basis

Is this the idea that content "just exists", the way that every number "exists" even if it's not referenced or stored anywhere? Like, the $10^{100}$ th prime number exists, even if we don't know what it is? I could see the justification in that. Otherwise it seems like a stretch.

A simple example is an ExternalHash, which is just 32 freely-chosen bytes which may or may not be even conceptually backed up by any content. If we want to say that that "hash" refers to the "inherently extant content" which may have never been generated and may never be known or even computable by any conscious entity in a finite universe, then, OK ;)

Reply:
I also notice you're using $\bigcup$ rather than $\bigsqcup$ -- is there ever a case where something could be part of both $E$ and $A$?

Reply:
Ah, I didn't know the meaning of $\bigsqcup$, but I looked it up, and yes that seems more appropriate